Allow volunteers to turn off intervention messages

eatyourgreens opened the pull request:
Staging branch URL: https://interventions.pfe-preview.zooniverse.org/

Fixes #4810.

Allow users to set `user.intervention_notifications` from their account settings page.
Add an opt-out checkbox to the Interventions component.
Track intervention messages and opt in as `classification.metadata.interventions.message` and `classification.metadata.interventions.opt_in`.
Adds tests for the Intervention component and intervention actions.

# Required Manual Testing

- [ ] Does the non-logged in home page render correctly?
- [ ] Does the logged in home page render correctly?
- [ ] Does the projects page render correctly?
- [ ] Can you load project home pages?
- [ ] Can you load the classification page?
- [ ] Can you submit a classification?
- [ ] Does talk load correctly?
- [ ] Can you post a talk comment?

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [ ] If the component is in coffeescript, is it converted to ES6? Is it free of eslint errors? Is the conversion its own commit?
- [ ] Are the tests passing locally and on Travis?

# Optional

- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you [resized and compressed](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) any image you've added?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?
- [ ] Have you followed the [Springer guidelines for commit messages](https://github.com/springernature/frontend-playbook/blob/master/git/git.md#commit-messages)?

eatyourgreens commented:
A heading might fix that. Messages are still unstyled too. Navigation is also clumsy if you get a message after clicking Done & Talk; you then have to click the Talk link again.

eatyourgreens commented:
Following on from Slack conversations, this PR should track whether or not a volunteer saw an intervention, and whether or not they opted out of receiving messages. The tracking info can be recorded on the classification, before it is saved to Panoptes.

eatyourgreens commented:
@camallen I've updated the settings form to use a fieldset and legend. I think that looks a lot better in terms of differentiating the notification preferences.
![Screenshot of the user settings form, showing the intervention settings with a legend and fieldset border.](https://user-images.githubusercontent.com/59547/45219008-6a6ff100-b2a1-11e8-9c1c-4df9c9641026.png)

eatyourgreens commented:
To try this out, send a message to a user on staging like this:
```shell
curl -vv \
     -X "POST" "https://notifications-staging.zooniverse.org/experiment" \
     -H 'Content-Type: application/json' \
     -u 'USERNAME:PASSWORD' \
     -d $'{
  "experiments": [{
    "message": "Hi, this is an intervention message.",
    "user_id": "1755",
    "delivered": false
  }]
}'
```
It should be shown to them the next time that they submit a classification, with an option to opt out of seeing these messages in future.

At the moment, there's no filtering to send messages to particular projects or workflows.

srallen commented:
@eatyourgreens is this still WIP? The PR has the WIP label still on it.

eatyourgreens commented:
@srallen functionality is mostly there but I don't think this is finished yet.

eatyourgreens commented:
I think this is good for review now.

wgranger commented:
I'm trying to test out this branch using the Curl command above with my username and password in place at `-u 'USERNAME:PASSWORD' \`, but I keep getting a 401, unauthorized when running from the command line.

wgranger commented:
Aaaa. That's what I'm missing. Let me give it another go.

wgranger commented on app/classifier/classifier.jsx:
```diff
@@ -262,35 +262,42 @@ class Classifier extends React.Component {
         height: innerHeight
+      },
+      'metadata.interventions': {
+        message: showIntervention,
```
Do we want message to always be a boolean here? If there isn't a user present I believe it would be set to `null`

eatyourgreens replied:
Good catch. I’ll update this and add some tests for the classification metadata.

eatyourgreens replied:
There's a test for this now, but the bug itself isn't fixed yet.

wgranger commented:
I got this working and posted an example of an intervention below. Aside from the one inline comment, my only pedantic suggestion is to add a margin underneath the dismissal box so it isn't touching the "Your classification:" header. 

![screen shot 2018-10-02 at 10 37 43 am](https://user-images.githubusercontent.com/14099077/46359970-033f2580-c630-11e8-8d90-7f290a31f0c5.png)

eatyourgreens commented:
Classification metadata should be fixed now, and I've added some basic styling to the bottom of the intervention message on the summary. I moved a lot of the classification code into Redux too, and added more tests.

eatyourgreens commented:
I also ended up rewriting `mockPanoptesResource` as a helper function for the tests.

eatyourgreens commented:
As an aside, I think the classifier is easier to debug in the Redux dev tools now.
![Screenshot of the Redux dev tools, showing the UPDATE_CLASSIFICATION and COMPLETE_CLASSIFICATION actions.](https://user-images.githubusercontent.com/59547/46422311-f9025180-c72b-11e8-8ab3-41cfbe94331f.png)

wgranger commented:
<img width="404" alt="screen shot 2018-10-03 at 1 43 24 pm" src="https://user-images.githubusercontent.com/14099077/46431809-62756680-c712-11e8-8981-002f078c7c1f.png">

Code looks good. Thanks for the test refactoring. Merging.
